### PR TITLE
Easymotion-like jump mode

### PR DESCRIPTION
**Pull request by Omnikar**
Depends on virtual text (#417). I will rebase once that's merged.

I've had to edit the functions for checking whether the primary cursor is in the view to accept any cursor position so they could be used to know when to stop finding jump locations.

To-do:
* [ ] Accept a motion to use for jumping (as of now, the `w` behaviour is used unconditionally).
* [ ] Change the color of the file text when in jump mode to make it easier to see the jump location annotations.

Related: #274, #510

### CODE REVIEW

File: helix-term/src/ui/mod.rs [hunk 1]
```diff
@@ -83,7 +83,11 @@ pub fn regex_prompt(
 
                             fun(view, doc, regex, event);
 
-                            view.ensure_cursor_in_view(doc, cx.editor.config.scrolloff);
+                            let cursor = doc
+                                .selection(view.id)
+                                .primary()
+                                .cursor(doc.text().slice(..));
+                            view.ensure_cursor_in_view(cursor, doc, cx.editor.config.scrolloff);
```
Comment (archseer) [inline helix-term/src/ui/mod.rs:90]: Why this change? You're calling it everywhere with the same primary cursor. Just fetch it inside the `ensure_cursor_in_view` method

Comment (Omnikar) [reply helix-term/src/ui/mod.rs:90]: I had to use the use the ensuring cursor in view functionality for locations other than that of the primary cursor in the jump mode functionality (when searching for jump locations, it has to stop once it reaches a location outside the screen). I suppose in order to avoid the resulting duplication in other locations that use this functionality, we could add a shortcut method, so you could do `doc.primary_cursor(view.id)`?

Comment (archseer) [reply helix-term/src/ui/mod.rs:90]: `is_cursor_in_view` seems to be used with other cursors, but I can't spot any ensure_ call that doesn't use the primary cursor

Comment (Omnikar) [reply helix-term/src/ui/mod.rs]: Ah

**Comment by Omnikar** [comment]
Replaced by #3791.
